cpp/include/raft/spatial/knn/detail/ball_cover/registers-inl.cuh

```diff
@@ -476,70 +477,112 @@ RAFT_KERNEL block_rbc_kernel_eps_dense(const value_t* X_index,
+  const int warp_id  = raft::shfl(threadIdx.x / WarpSize, 0);
+  const int query_id = raft::shfl(blockIdx.x * num_warps + warp_id, 0);
 
-  for (value_int cur_k = 0; cur_k < n_landmarks; ++cur_k) {
-    // TODO: this might also be worth computing in-place here
-    value_t cur_R_dist = R_dists[blockIdx.x * n_landmarks + cur_k];
+  // this is an early out for a full warp
+  if (query_id >= n_queries) return;
 
-    // prune all R's that can't be within eps
-    if (cur_R_dist - R_radius[cur_k] > eps) continue;
+  unsigned long long int column_count = 0;
 
-    // The whole warp should iterate through the elements in the current R
-    value_idx R_start_offset = R_indptr[cur_k];
-    value_idx R_stop_offset  = R_indptr[cur_k + 1];
-
-    value_idx R_size = R_stop_offset - R_start_offset;
+  const value_t* x_ptr = X + (n_cols * query_id);
 
-    value_int limit = Pow2<WarpSize>::roundDown(R_size);
-    value_int i     = threadIdx.x;
-    for (; i < limit; i += tpb) {
-      // Index and distance of current candidate's nearest landmark
-      value_idx cur_candidate_ind = R_1nn_cols[R_start_offset + i];
-      value_t cur_candidate_dist  = R_1nn_dists[R_start_offset + i];
+  // we omit the sqrt() in the inner distance compute
+  const value_t eps2 = eps * eps;
 
-      const value_t* y_ptr = X_index + (n_cols * cur_candidate_ind);
-      if (dfunc(x_ptr, y_ptr, n_cols) <= eps) {
-        adj[blockIdx.x * m + cur_candidate_ind] = true;
-        if (vd != nullptr) atomicAdd(&column_count_smem, 1);
+#pragma nounroll
+  for (uint32_t cur_k0 = 0; cur_k0 < n_landmarks; cur_k0 += WarpSize) {
+    // Pre-compute landmark_dist & triangularization checks for 32 iterations
+    // prune all R's that can't be within eps
+    const uint32_t lane_k     = cur_k0 + lid;
+    const value_t lane_R_dist = lane_k < n_landmarks
+                                  ? raft::sqrt(dfunc(x_ptr, R + lane_k * n_cols, n_cols))
+                                  : std::numeric_limits<value_idx>::max();
+    const int lane_check =
+      lane_k < n_landmarks ? static_cast<int>(lane_R_dist - R_radius[lane_k] <= eps) : 0;
+
+    int lane_mask = raft::ballot(lane_check);
+    if (lane_mask == 0) continue;
+
+    uint32_t k_offset = __ffs(lane_mask) - 1;
+    do {
+      const uint32_t cur_k = cur_k0 + k_offset;
+
+      // update lane_mask for next iteration - erase bits up to k_offset
+      lane_mask &= -(1 << k_offset + 1);
+
+      // The whole warp should iterate through the elements in the current R
+      const value_idx R_start_offset = R_indptr[cur_k];
+      const uint32_t R_size          = R_indptr[cur_k + 1] - R_start_offset;
+
+      // we have precomputed the query<->landmark distance
+      const value_t cur_R_dist = raft::shfl(lane_R_dist, k_offset);
+
+      const uint32_t limit = Pow2<WarpSize>::roundDown(R_size);
+      int i                = limit + lid;
+
+      // look ahead for next k_offset
+      k_offset = lane_mask != 0 ? __ffs(lane_mask) - 1 : WarpSize;
```

Comment:
To save a few instructions, I would remove the ternary expression here:

```c++
k_offset = __ffs(lane_mask) - 1;
```

And change the loop condition to:

```c++
while (lane_mask);
```

But then to avoid the unnecessary invocation when the mask is 0, the `__ffs` would be moved to the top of the loop instead of computing ahead for the next iteration. It would look like:

```c++
int lane_mask = raft::ballot(lane_check);

while(lane_mask) {
    uint32_t k_offset = __ffs(lane_mask) - 1;
    ...
}
```

Also, as discussed offline, there is an instruction to find the most significant bit but not the least, so you can change the order to avoid the bit reversal:

```c++
uint32_t k_offset = 31 - __clz(lane_mask);
```

Reply:
As a side note, if you need to keep the previous order from the least significant 1, you can use `__brev` before the loop and then `__clz` inside the loop. I think that the current version has to do the bit reverse every iteration because the mask is changed.